### PR TITLE
Changling evolution menu becomes cellular emporium

### PR DESCRIPTION
**Pull request by coiax**
Now with TGUI! Wow!

:cl: coiax
rscadd: Changelings now evolve powers via the cellular emporium, which they can access via an action button.
/:cl:

image

It is pretty bare bones and basic, but it works.

And still a net reduction in lines of code, and blah blah MVC and all that.

Moving to this means that the layout can be prettified without having to change the code, same with appearance, and so on. TOTAL SEPERATION etc etc.
- All changeling powers are still activated and used in the same way, this is just the storefront.
- Adds an "always" state for tgui, which means you'll always be able to use it. Additional restrictions on the evolution of powers and whatnot are still done in code, but not even death stops a changeling from evolving various powers, so needed a new state type.
## Would be nice but not getting in the way of this getting merged (because it would be helpful for datum antags, see?)
- A different template/ui theme (like the difference between NT and Syndicate)
- A better layout by someone who can actually do stuff

### CODE REVIEW

File: code/game/gamemodes/changeling/cellular_emporium.dm [hunk 1]
```diff
@@ -0,0 +1,81 @@
+// cellular emporium
+// The place where changelings go to buy their biological weaponry.
+
+/datum/cellular_emporium
+	var/name = "cellular emporium"
+	var/datum/changeling/changeling
+
+/datum/cellular_emporium/New(my_changeling)
+	changeling = my_changeling
```
Comment (RemieRichards) [inline code/game/gamemodes/changeling/cellular_emporium.dm]: call ```..()```

File: code/game/gamemodes/changeling/cellular_emporium.dm [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	var/absorbed_dna_count = changeling.absorbedcount
+
+	data["can_readapt"] = can_readapt
+	data["genetic_points_remaining"] = genetic_points_remaining
+	data["absorbed_dna_count"] = absorbed_dna_count
+
+	var/list/abilities = list()
+
+	for(var/path in subtypesof(/obj/effect/proc_holder/changeling))
+		var/obj/effect/proc_holder/changeling/ability = new path()
```
Comment (RemieRichards) [inline code/game/gamemodes/changeling/cellular_emporium.dm]: No, this is not how we access vars from paths.

Comment (coiax) [inline code/game/gamemodes/changeling/cellular_emporium.dm]: You are correct, I'll get it initial()'d. (I was just copying what the old code did, honest)

File: tgui/src/interfaces/cellular_emporium.ract [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  <ui-section label='{{name}}' candystripe right>
+    <span>{{desc}}</span>
+    <span>{{helptext}}</span>
+    <span>Cost: {{dna_cost}}</span>
+
+    {{#if required_absorbtions}}
+      <span>Required Absorbtions: {{required_absorbtions}}
+    {{/if}}
+
+    <ui-button state='{{owned ? "selected" : can_purchase ? null : "disabled"}}'
```
Comment (RemieRichards) [inline tgui/src/interfaces/cellular_emporium.ract]: I feel this should be on a new line, so it's in a consistent place.

File: code/game/gamemodes/changeling/cellular_emporium.dm [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+		var/dna_cost = initial(ability.dna_cost)
+		if(dna_cost <= 0)
+			continue
+
+		var/list/AL = list()
+		AL["name"] = initial(ability.name)
+		AL["desc"] = initial(ability.desc)
+		AL["helptext"] = initial(ability.helptext)
+		AL["owned"] = changeling.has_sting(ability)
```
Comment (RemieRichards) [inline code/game/gamemodes/changeling/cellular_emporium.dm]: ```has_sting()``` is ass and doesn't work on paths, that will need fixing.

**Comment by xxalpha** [comment]
Could you put those buttons in a column or something

**Comment by Incoming5643** [comment]
Cursory dismissal of shoving Tgui into things

**Comment by RemieRichards** [comment]
Dismissal of having a computer from the 1800s

**Comment by xxalpha** [comment]
If you make a web page that takes so long to load and it is nothing but a table of text and hyperlinks, you fucked up somewhere.

File: code/game/gamemodes/changeling/cellular_emporium.dm (hunk 4 above)
Comment (coiax) [inline code/game/gamemodes/changeling/cellular_emporium.dm]: I fixed that.

**Comment by PKPenguin321** [comment]
The Evolve buttons should be to the left of the powers name so that they're not all over the place

**Comment by TechnoAlchemisto** [comment]
What PKP said, evolve buttons should all be in a line, not at the end of the fluff.

**Comment by RemieRichards** [comment]
Omg you guys, I pointed that out, and it was fixed, 5 HOURS ago.

**Comment by xxalpha** [comment]
Excuse me, I pointed it out first to put it in a dedicated column.

**Comment by TechnoAlchemisto** [comment]
hey man not my fault the op didnt update the image shit

**Comment by coiax** [comment]
I haven't put the buttons in a column?

**Comment by Durkel** [comment]
Even though it shouldn't break anything at all, make sure you test it so the respecing works.

**Comment by coiax** [comment]
I have checked that you can readapt after husking a corpse and it works as normal.

**Comment by lzimann** [comment]
as much as I like tgui, the bigger ones lag quite a lot, i'm not sure if it's good to update everything to it

**Comment by coiax** [comment]
From a code perspective, the old evolution menu is SUPER UGLY. I regret nothing.

**Comment by xxalpha** [comment]
The current UI for it is awful, yes.

**Comment by RemieRichards** [comment]
@lzimann this tgui isn't sending much data, it should be fine.

**Comment by Razharas** [comment]
I dont like how much space tgui takes

**Comment by AnturK** [comment]
Make the powers descriptions extendable like uplink items maybe ? Put up fresh image if you changed the layout please. (and merge conflicts)

**Comment by Supermichael777** [comment]
old menu was cancer too easy to get wrong power

**Comment by optimumtact** [comment]
Is this ready?

**Comment by TechnoAlchemisto** [comment]
evolve buttons need to be collumnized

**Comment by coiax** [comment]
Thanks to the separation of form from function that's easy to do for someone who knows how to make pretty HTML things. I am not one of those people.

Can we get this merged soon, it's part of my datum antags roadmap

**Pull request merged**